src/urbit.ts

```diff
@@ -0,0 +1,468 @@
+          } else {
+            console.log('Unrecognized response', data);
+          }
+          // An incoming message, for example:
+          // {
+          //   id: 10,
+          //   json: {
+          //     'chat-update' : { // This is where we hook our "on" handlers like "onChatUpdate"
+          //       message: {
+          //         envelope: {
+          //           author: 'zod',
+          //           letter: {
+          //             text: 'hi'
+          //           },
+          //           number: 10,
+          //           uid: 'saludhafhsdf',
+          //           when: 124459
+          //         },
+          //         path: '/~zod/mailbox'
+          //       }
+          //     }
+          //   }
+          // }
+        }
+      });
+      this.sseClient?.addEventListener('error', function(event: Event) {
+        console.error('pipe error', event);
+      });
+    }
+    return this.sseClient;
+  }
+
+  /**
+   * Autoincrements the next event ID for the appropriate channel.
+   */
+  getEventId(): number {
+    this.lastEventId = Number(this.lastEventId) + 1;
+    return this.lastEventId;
+  }
+
+  /**
+   * Acknowledges an event.
+   *
+   * @param eventId The event to acknowledge.
+   */
+  async ack(eventId: number): Promise<void | number> {
+    return this.sendMessage('ack', { 'event-id': eventId });
+  }
+
+  /**
+   * This is a wrapper method that can be used to send any action with data.
+   *
+   * Every message sent has some common parameters, like method, headers, and data
+   * structure, so this method exists to prevent duplication.
+   *
+   * @param action The action to send
+   * @param data The data to send with the action
+   * 
+   * @returns void | number If successful, returns the number of the message that was sent
+   */
+  async sendMessage(action: string, data?: object): Promise<void | number> {
+    const headers: headers = {
+      'Content-Type': 'application/json',
+    };
+    if (!isBrowser) {
+      headers.Cookie = this.cookie;
+    }
+    const id = this.getEventId();
+    if (this.verbose) {
+      console.log(`Sending message ${id}:`, action, data, headers);
+    }
+    return fetch(this.channelUrl, {
+      method: 'put',
+      headers,
+      credentials: 'include',
+      body: JSON.stringify([{
+        id,
+        action,
+        ...data,
+      }])
+    }).catch((error: any) => {
+      console.error('message error', error);
+    })
+    .then(response => {
+      if (this.verbose) {
+        console.log(`Received from message ${id}: `, response);
+      }
+      return id;
+    });
+  }
+
+  /**
+   * Pokes a ship with data.
+   *
+   * @param ship The ship to poke
+   * @param app The app to poke
+   * @param mark The mark of the data being sent
+   * @param json The data to send
+   */
+  async poke(
+    ship: string = this.ship ? this.ship : '',
+    app: string,
+    mark: string,
+    json: Object,
+    successFunc = () => {},
```

Comment:
Ok so cross-ship pokes from eyre are heavily discouraged and don't work when you need a mark translation (i.e. you can only do a cross-ship poke if the mark is `'json'`). I'd make ship mandatory in the constructor and drop it from this function. Furthermore, (imo) you should drop the callbacks and just use a Promise that resolves when the poke-acks

Reply:
<img width="516" alt="image" src="https://user-images.githubusercontent.com/3999320/98836010-05226800-23f6-11eb-8b9f-23718fe2771f.png">
Do you have any problem with this — setting the ship when we receive the cookie back?